core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```diff
@@ -379,19 +378,30 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
         // register the user
         User user = createAccount(si.username,si.password1);
         user.setFullName(si.fullname);
-        try {
-            // legacy hack. mail support has moved out to a separate plugin
-            Class<?> up = Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
-            Constructor<?> c = up.getDeclaredConstructor(String.class);
-            user.addProperty((UserProperty)c.newInstance(si.email));
-        } catch (RuntimeException e) {
-            throw e;
-        } catch (Exception e) {
-            LOGGER.log(Level.WARNING, "Failed to set the e-mail address",e);
+        if(isMailerPluginPresent()) {
+            try {
+                // legacy hack. mail support has moved out to a separate plugin
+                Class<?> up = Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
+                Constructor<?> c = up.getDeclaredConstructor(String.class);
+                user.addProperty((UserProperty)c.newInstance(si.email));
+            } catch (ReflectiveOperationException e) {
+                throw new RuntimeException(e);
+            }
```

Comment:
It's not attempting anything, it doesn't proceed into this block unless isMailerPluginPresent() is true, in which case the class check is already done and instead of silently failing as it was doing, it will throw an exception.